ARROW-9902: [Rust] [DataFusion] Add array() built-in function

jorgecarleitao opened the pull request:
This adds `array()` built-in function to most primitive types. For composite types, this is more challenging and I decided to scope out of this PR.

andygrove commented on rust/datafusion/src/physical_plan/array_expressions.rs:
```diff
@@ -0,0 +1,102 @@
+                    builder.values().append_value(arg.value(index))?;
+                }
+            }
+            builder.append(true).unwrap();
```
Can we remove this unwrap?

andygrove commented:
@jorgecarleitao please rebase

andygrove commented on rust/datafusion/src/physical_plan/array_expressions.rs:
```diff
@@ -0,0 +1,102 @@
+        DataType::Float32 => array!(args, Float32Array, Float32Builder),
+        DataType::Float64 => array!(args, Float64Array, Float64Builder),
+        DataType::Int8 => array!(args, Int8Array, Int8Builder),
+        DataType::Int32 => array!(args, Int32Array, Int32Builder),
```
Is it intentional that 16 bit integer types are not represented here?

jhorstmann commented:
@jorgecarleitao do you already have followup plans for more array functions? some functions like `array_first`, `array_last`, `array_index_of or `array_contains` would be useful for me too.

andygrove commented on rust/datafusion/src/physical_plan/array_expressions.rs:
```diff
@@ -0,0 +1,106 @@
+        ));
+    }
+
+    match args[0].data_type() {
```
It looks like we assume all values in the list are going to be the same type, which makes sense, but I don't see any validation checks for that. Perhaps we need to add them, or do we expect that type coercion will take care of introducing the appropriate casts? Either way, a specific test for this would be good.

jorgecarleitao replied:
Type coercion should do it, but while adding a test to this, I found a bug in the coercion logic. So, this PR will have to wait a bit until the logic is fixed. Spot-on, @andygrove !

jorgecarleitao commented:
This is pending #8108, as it requires a bug fix to pass the tests.

jorgecarleitao commented:
@jhorstmann , those are great suggestions!

I am not planning to work on that atm, as I wanted first to have an interface to declare UDFs with the same capabilities as built-in functions (#7967 and #8032).

However, please add them to the [issue tracker](https://issues.apache.org/jira/issues/?jql=project%20%3D%20ARROW%20AND%20resolution%20%3D%20Unresolved%20AND%20component%20in%20(Rust%2C%20%22Rust%20-%20DataFusion%22)%20ORDER%20BY%20created%20DESC%2C%20updated%20ASC%2C%20priority%20DESC). You are of course also welcome to take them, if you have the time. They seem to be good entry-level issues, but of course the devil is always in the details ^_^

andygrove commented:
@jorgecarleitao I plan on reviewing this one tomorrow.

andygrove commented on rust/datafusion/tests/sql.rs:
```diff
@@ -734,7 +737,38 @@ fn query_concat() -> Result<()> {
+    let mut ctx = ExecutionContext::new();
+    ctx.register_table("test", Box::new(table));
+    let sql = "SELECT array(c1, cast(c2 as varchar)) FROM test";
+    let actual = execute(&mut ctx, sql);
```
nit: It would be good to have an assertion here looking at the data type of this array

jorgecarleitao replied:
Yeah, we have not been testing that the final schema is expected in any of these tests, and IMO we should, for all tests in this module. I will open an issue in Jira with this.

jorgecarleitao replied:
Filed in ARROW-10036

andygrove commented on rust/datafusion/tests/sql.rs:
```diff
@@ -609,6 +609,74 @@ fn execute(ctx: &mut ExecutionContext, sql: &str) -> Vec<String> {
+
+            let mut r = Vec::with_capacity(*n as usize);
+            for i in 0..*n {
+                let array = array.value(row_index);
```
I think this also needs null handling since elements within this array could have null values?

jorgecarleitao replied:
Not sure I follow:

* if `array.is_null(row_index)`, then line 613 covers it with `array.is_null(row_index)`
* if the elements within the array are null, then line 613 covers it because we call the function recursively on each element, in line 672.

I moved that statement away from the loop to make it more explicit.

andygrove reviewed (approved):
LGTM with a couple minor comments

andygrove commented on rust/datafusion/tests/sql.rs:
```diff
@@ -609,6 +609,74 @@ fn execute(ctx: &mut ExecutionContext, sql: &str) -> Vec<String> {
+        }
+        DataType::FixedSizeList(_, n) => {
+            let array = array.as_any().downcast_ref::<FixedSizeListArray>().unwrap();
+            let array = array.value(row_index);
```
I think that we need a null check here in case the whole array is null. I know we check at the start of this method for a null value _within_ an array, but I don't see a check for a null array. Maybe there are guarantees that no arrays can be null? If so, it would be good to add a comment here to explain.

```rust
let array = array.as_any().downcast_ref::<FixedSizeListArray>().unwrap();
if array.is_null(row_index) {
  // this case isn't handled
} else {
  let array = array.value(row_index);
  let mut r = Vec::with_capacity(*n as usize);
  for i in 0..*n {
    r.push(array_str(&array, i as usize));
  }
  format!("[{}]", r.join(","))
}
```

jorgecarleitao commented on rust/datafusion/tests/sql.rs:
```diff
@@ -609,6 +609,74 @@ fn execute(ctx: &mut ExecutionContext, sql: &str) -> Vec<String> {
 }
 
+fn array_str(array: &Arc<dyn Array>, row_index: usize) -> String {
+    if array.is_null(row_index) {
+        return "NULL".to_string();
+    }
```
@andygrove , isn't this the check that you refer to? I moved it to the beggining of the function so that it applies to all types (array or not).

andygrove replied:
This check is good but no, this is not what I'm referring to. Here is my example code again but renaming some variables for clarity.

```rust
let array1 = array.as_any().downcast_ref::<FixedSizeListArray>().unwrap();
if array1.is_null(row_index) {
  // this case isn't handled
} else {
  let array2 = array1.value(row_index);
  let mut r = Vec::with_capacity(*n as usize);
  for i in 0..*n {
    r.push(array_str(&array2, i as usize));
  }
  format!("[{}]", r.join(","))
}
```

There are two arrays involved here; array1 and array2.

`array1` is the array of columns.

`array2` is one of the columns from the array `array2`.

We call `array_str` with `array2` and that part is all good.

However, when we initialize `array2` with `let array2 = array1.value(row_index);` we are not first checking to see if there is a valid value at `row_index` in `array1`.

Does that make sense?

jorgecarleitao commented:
@andygrove , I rebased and added a comment to improve readability of the `array_str`. Note that this conflicts with the Min/Max of a string, #8215 . Thank you for your patience here 🥇

andygrove commented:
> @andygrove , I rebased and added a comment to improve readability of the `array_str`. Note that this conflicts with the Min/Max of a string, #8215 . Thank you for your patience here 1st_place_medal

Oh ... no, this is my bad, I was misreading the code. The comment helped though. Thank you!
